include/ck/utility/amd_wmma.hpp

```diff
@@ -0,0 +1,136 @@
+
+// wave32 only
+// src: fp16, dst: fp32
+template <index_t MPerWave, index_t NPerWave>
+struct intrin_wmma_f32_16x16x16_f16_w32;
+
+template <>
+struct intrin_wmma_f32_16x16x16_f16_w32<16, 16>
+{
+    template <class FloatC>
+    __device__ static void Run(const half16_t& reg_a, const half16_t& reg_b, FloatC& reg_c)
+    {
+        reg_c.template AsType<float8_t>()(Number<0>{}) = __builtin_amdgcn_wmma_f32_16x16x16_f16_w32(
+            reg_a, reg_b, reg_c.template AsType<float8_t>()[Number<0>{}]);
+    }
+};
+
+// src: bf16, dst: fp32
+template <index_t MPerWave, index_t NPerWave>
+struct intrin_wmma_f32_16x16x16_bf16_w32;
+
+template <>
+struct intrin_wmma_f32_16x16x16_bf16_w32<16, 16>
+{
+    template <class FloatC>
+    __device__ static void Run(const bhalf16_t& reg_a, const bhalf16_t& reg_b, FloatC& reg_c)
+    {
+        reg_c.template AsType<float8_t>()(Number<0>{}) =
+            __builtin_amdgcn_wmma_f32_16x16x16_bf16_w32(
+                reg_a, reg_b, reg_c.template AsType<float8_t>()[Number<0>{}]);
+    }
+};
+
+// src: fp16, dst: fp16
+template <index_t MPerWave, index_t NPerWave>
+struct intrin_wmma_f16_16x16x16_f16_w32;
+
+template <>
+struct intrin_wmma_f16_16x16x16_f16_w32<16, 16>
+{
+    template <class FloatC>
+    __device__ static void
+    Run(const half16_t& reg_a, const half16_t& reg_b, FloatC& reg_c, const bool opsel)
+    {
+        // opsel usage
+        // false: D0.[0:15] = result
+        // true : D0.[16:31]= result
+        reg_c.template AsType<half16_t>()(Number<0>{}) = __builtin_amdgcn_wmma_f16_16x16x16_f16_w32(
+            reg_a, reg_b, reg_c.template AsType<half16_t>()[Number<0>{}], opsel);
+    }
+};
+
+// src: bf16, dst: bf32
+template <index_t MPerWave, index_t NPerWave>
+struct intrin_wmma_bf16_16x16x16_bf16_w32;
+
+template <>
+struct intrin_wmma_bf16_16x16x16_bf16_w32<16, 16>
+{
+    template <class FloatC>
+    __device__ static void
+    Run(const bhalf16_t& reg_a, const bhalf16_t& reg_b, FloatC& reg_c, const bool opsel)
+    {
+        // opsel usage
+        // false: D0.[0:15] = result
+        // true : D0.[16:31]= result
+        reg_c.template AsType<bhalf16_t>()(Number<0>{}) =
+            __builtin_amdgcn_wmma_bf16_16x16x16_bf16_w32(
+                reg_a, reg_b, reg_c.template AsType<bhalf16_t>()[Number<0>{}], opsel);
+    }
+};
+
+// src: iu8, dst: i32
+template <index_t MPerWave, index_t NPerWave>
+struct intrin_wmma_i32_16x16x16_iu8_w32;
+
+template <>
+struct intrin_wmma_i32_16x16x16_iu8_w32<16, 16>
+{
+    template <class FloatC>
+    __device__ static void Run(const bool neg_a,
+                               const int8x16_t& reg_a,
+                               const bool neg_b,
+                               const int8x16_t& reg_b,
+                               FloatC& reg_c,
+                               const bool clamp)
+    {
+        reg_c.template AsType<int32x8_t>()(Number<0>{}) =
+            __builtin_amdgcn_wmma_i32_16x16x16_iu8_w32(
+                neg_a,
```

Comment:
Can you pass `neg_a` and `neg_b` as function arguments? I suppose they are immediate values (ie part of instruction encoding) therefore can only be passed as template arguments.

Is the wrapper `intrin_wmma_i32_16x16x16_iu8_w32` tested somewhere? Could you confirm?

Reply:
```c++
    template <class FloatC>
    __device__ static void Run(const bool neg_a,
                               const int8x16_t& reg_a,
                               const bool neg_b,
                               const int8x16_t& reg_b,
                               FloatC& reg_c,
                               const bool clamp)
```

I am referring to the arguments `const bool neg_a` and `const bool neg_b` of the wrapper above.

Not `__builtin_amdgcn_wmma_i32_16x16x16_iu8_w32` itself.

Reply:
https://github.com/llvm/llvm-project/blob/f59f116bd5c357b1cb4b04693c88d41484e168d5/llvm/include/llvm/IR/IntrinsicsAMDGPU.td#L2048-L2060

```
class AMDGPUWmmaIntrinsicIU<LLVMType AB, LLVMType CD> :
  Intrinsic<
    [CD],               // %D
    [
      llvm_i1_ty,       // %A_sign
      AB,               // %A
      llvm_i1_ty,       // %B_sign
      AB,               // %B
      LLVMMatchType<0>, // %C
      llvm_i1_ty,       // %clamp
    ],
    [IntrNoMem, IntrConvergent, IntrWillReturn, ImmArg<ArgIndex<0>>, ImmArg<ArgIndex<2>>, ImmArg<ArgIndex<5>>]
>;
```

The definition in LLVM repo states that `a_sign`, `b_sign`, `clamp` are immediate arguments.

Reply:
Hi @rosenrodt , it's a good idea to add unit test cover all the builtin function, I'll do it.
However, I'm not sure about what you talk about "immediate arguments", my current wrapper pass them as function arguments with const prefix, do you want me modify them to non-type template arguments?

Reply:
@aska-0096 My bad. I was under the impression that immediate values need to be compile-time constant. That I think as long as you write a test to simply invoke the wrapper and it passes compilation then it is good.

Reply:
Hi @rosenrodt 
Verified your previous thought that the immediate value should be compile-time constant during the implementation of unit test, great!